[#13] Redis 세션 스토리지 연동

cyj199637 opened the pull request:
- Redis를 사용하기 위한 의존성 추가
- 로컬에 설치된 Redis와 연결하기 위한 RedisConfig 추가

cyj199637 committed:
[#13] Redis 세션 스토리지 연동

- Redis를 사용하기 위한 의존성 추가
- 로컬에 설치된 Redis와 연결하기 위한 RedisConfig 추가

f-lab-dev commented on src/main/java/me/liiot/snsserver/model/User.java:
```diff
@@ -19,7 +20,7 @@
 @Builder
 @AllArgsConstructor
-public class User {
+public class User implements Serializable {
```
1. 일반적으로 자바 직렬화를 권장하지 않습니다. 왜일까요? (spring document를 뒤져보면 나옵니다~)
2. JSON 직렬화를 사용해볼 수 있지 않을까요?

cyj199637 replied:
자바 직렬화는 비정상적인 접근을 통해 확인되지 않은 바이트 코드가 입력된 페이로드를 통해 원격 명령이 실행된다는 취약점이 있다. 이렇게 조작된 입력 데이터가 역직렬화 과정을 거친다면 애플리케이션이 제대로 동작하지 않을 수 있다. 따라서, 안전하지 않은 환경에서는 자바 직렬화를 권장하지 않고 있다.

cyj199637 committed:
Merge branch 'feature/11' into feature/13

cyj199637 committed:
[#13] 피드백 반영

- User 객체가 JSON 직렬화가 되도록 수정

f-lab-dev commented on src/main/java/me/liiot/snsserver/aspect/CheckLoginAspect.java:
```diff
@@ -21,7 +23,15 @@
 
-        User currentUser = (User)httpSession.getAttribute(SessionKeys.USER);
+        User currentUser = null;
+        ObjectMapper mapper = new ObjectMapper();
```
이 ObjectMapper는 thread safe하기에 한번만 생성해서 돌려쓸 수 있습니다~ 굳이 매번 생성해서 객체생성에 따른 부하를 만들지 않는게 좋아보입니다

f-lab-dev reviewed (commented):
개선해주신 코드에서는 직렬화를 직접 하고 계신데요, 스프링에서 이런것들을 지원합니다. 한번 찾아보시면 좋을 것 같네요~

cyj199637 committed:
[#13] 피드백 반영

- ObjectMapper를 Bean으로 등록하여 한번만 생성되도록 수정

cyj199637 committed:
[#13] 피드백 반영

- ObjectMapper를 빈으로 등록

f-lab-dev commented on src/main/java/me/liiot/snsserver/config/RedisConfig.java:
```diff
@@ -16,6 +17,11 @@
 
+    @Bean
+    public ObjectMapper objectMapper() {
+        return new ObjectMapper();
```
Spring에서는 `ObjectMapper`가 기본적으로 bean에 등록되어있습니다~
- 그 bean은 어디어디에 쓰일까요? 
- 그 bean은 이 bean의 목적과 동일할까요? 즉 옵션이 같을까요?

cyj199637 replied:
* Spring에서는 요청 본문에서 메시지를 읽어들이거나 응답 본문에 메시지를 작성할 때 HttpMessageConverters를 사용한다. HttpMessageConverters는 요청 본문에서 읽어들인 문자열을 자바 오브젝트로 변환하거나 바인딩 오브젝트를 JSON이나 XML 등으로 변환한다. Spring에는 기본적으로 여러 종류의 HttpMessageConverters들이 등록되어있는데 그 중 MappingJacksonHttpMessageConverter와 MappingJackson2HttpMessageConverter2가 Jackson ObjectMapper를 이용하여 JSON과 자바 오브젝트 간 변환 작업을 수행한다.

* 위의 ObjectMapper도 외부 세션 스토리지로 자바 오브젝트를 보내기 위해 JSON으로 변환하고 이를 다시 읽어들여 자바 오브젝트로 변환하는데 사용된다는 점에서 HttpMessageConverter의 ObjectMapper와 목적이 동일하다고 생각합니다.

cyj199637 committed:
[#13] 피드백 반영

- ObjectMapper 대신 GenericJackson2JsonRedisSerializer로 직렬화

f-lab-dev commented on src/main/java/me/liiot/snsserver/aspect/CheckLoginAspect.java:
```diff
@@ -20,19 +19,13 @@
-            throw new RuntimeException("로그인 인증을 하는 과정에서 에러가 발생하였습니다.", e);
-        }
+        byte[] serializedSource = (byte[]) httpSession.getAttribute(SessionKeys.USER);
+        User currentUser = serializer.deserialize(serializedSource, User.class);
```
이 메소드는 로그인을 했는지 확인하는 메소드인데 **세션에서 가져오는 데이터를 역직렬화한다**라는 역할까지 맡고 있습니다.
클래스나 메소드의 역할과 다른 역할 같은데요~ 세션에서 로그인관련 데이터를 입출력하는 클래스를 따로 분리하면 역할이 분리되지 않을까요?

f-lab-dev commented on src/main/java/me/liiot/snsserver/config/RedisConfig.java:
```diff
@@ -18,8 +18,8 @@
-    public ObjectMapper objectMapper() {
-        return new ObjectMapper();
+    public GenericJackson2JsonRedisSerializer genericJackson2JsonRedisSerializer() {
+        return new GenericJackson2JsonRedisSerializer();
```
이런 serializer도 레디스 세션 serializer bean으로 등록해두면 세션이 알아서 직렬화/역직렬화를 해줍니다.
이 부분은 document를 좀 잘 뒤져봐야 나올 것 같긴 하네요.
`spring redis session serializer` 이런 키워드로 검색해보시면 나올 것 같아요~

cyj199637 committed:
[#13] RedisTemplate 설정

cyj199637 committed:
[#13] 피드백 반영

- CheckLoginAspect와 CurrentUserArgumentResolver에서 HttpSession에 관한 로직을 직접 다루지 않도록 분리
- Session에 User객체의 Id만 저장하도록 변경
- 위의 사항으로 인해 영향을 받는 사용자 정보 업데이트 기능 로직 변경

f-lab-dev commented on src/main/java/me/liiot/snsserver/service/SessionLoginService.java:
```diff
@@ -13,16 +12,20 @@
-    private final GenericJackson2JsonRedisSerializer serializer;
-
     @Override
     public void loginUser(User user) {
```
만약 `userId`만 사용한다고 하면 long 타입의 userId만 인자로 받아도 될 것 같네요. 그럼 `User`클래스와의 의존성도 끊을 수 있을 것 같아요~

f-lab-dev commented on src/main/java/me/liiot/snsserver/config/RedisConfig.java:
```diff
@@ -30,4 +32,13 @@ RedisConnectionFactory redisConnectionFactory() {
     }
+
+    @Bean
+    RedisTemplate<String, Object> redisTemplate() {
```
전체 공통으로 사용되는 `redisTemplate`을 재정의하신건데 이러면 다른 영향은 없을지 확인해보셨나요?

cyj199637 replied:
네! RedisTemplate이 객체를 직렬화/역직렬화하는데 사용하는데 default로 사용되는 serializer가 JdkSerializationRedisSerializer라고 공식 document에 쓰여져 있었습니다. JdkSerializationRedisSerializer로
직렬화하면 바이너리 값으로 저장되기 때문에 조회하는데 어려움이 있다고 판단했습니다. 그래서
key는 StringRedisSerializer로 value는 json 형태로 저장하기 위해 GenericJackson2JsonRedisSerializer를
사용하도록 재정의했습니다.

cyj199637 committed:
[#13] 피드백 반영

- LoginServie의 loginUser에서 User 객체 대신 userId 값만 받도록 수정
